src/optimize.jl

```diff
@@ -0,0 +1,118 @@
+
+# Arguments
+- `objective`: Variational Objective.
+- `λ₀`: Initial value of the variational parameters.
+- `restruct`: Function that reconstructs the variational approximation from the flattened parameters.
+- `q`: Initial variational approximation. The variational parameters must be extractable through `Optimisers.destructure`.
+- `n_max_iter`: Maximum number of iterations.
+- `objargs...`: Arguments to be passed to `objective`.
+- `kwargs...`: Additional keywoard arguments. (See below.)
+
+# Keyword Arguments
+- `adbackend`: Automatic differentiation backend. (Type: `<: ADtypes.AbstractADType`.)
+- `optimizer`: Optimizer used for inference. (Type: `<: Optimisers.AbstractRule`; Default: `Adam`.)
+- `rng`: Random number generator. (Type: `<: AbstractRNG`; Default: `Random.default_rng()`.)
+- `show_progress`: Whether to show the progress bar. (Type: `<: Bool`; Default: `true`.)
+- `callback!`: Callback function called after every iteration. The signature is `cb(; stats, restructure, λ, g)`, which returns a dictionary-like object containing statistics to be displayed on the progress bar. The variational approximation can be reconstructed as `restructure(λ)`, `g` is the stochastic estimate of the gradient. (Default: `nothing`.)
+- `prog`: Progress bar configuration. (Default: `ProgressMeter.Progress(n_max_iter; desc="Optimizing", barlen=31, showspeed=true, enabled=prog)`.)
+- `state`: Initial value for the internal state of optimization. Used to warm-start from the state of a previous run. (See the returned values below.) (Type: `<: NamedTuple`.)
+
+# Returns
+- `λ`: Variational parameters optimizing the variational objective.
+- `logstats`: Statistics and logs gathered during optimization.
+- `states`: Collection of the final internal states of optimization. This can used later to warm-start from the last iteration of the corresponding run.
+"""
+function optimize(
+    prob,
+    objective    ::AbstractVariationalObjective,
+    restructure,
+    λ₀           ::AbstractVector{<:Real},
+    n_max_iter   ::Int,
+    objargs...;
+    adbackend    ::ADTypes.AbstractADType, 
+    optimizer    ::Optimisers.AbstractRule = Optimisers.Adam(),
+    rng          ::Random.AbstractRNG      = Random.default_rng(),
+    show_progress::Bool                    = true,
+    state        ::NamedTuple              = NamedTuple(),
+    callback!                              = nothing,
+    prog                                   = ProgressMeter.Progress(
+        n_max_iter;
+        desc      = "Optimizing",
+        barlen    = 31,
+        showspeed = true,
+        enabled   = show_progress
+    )
+)
+    λ        = copy(λ₀)
+    opt_st   = haskey(state, :opt) ? state.opt : Optimisers.setup(optimizer, λ)
+    obj_st   = haskey(state, :obj) ? state.obj : init(rng, objective, λ, restructure)
```

Comment:
Can we hide this behind a `get_or_init(state)` or something? 

And is this `state` to be shared across different methods, or are we expecting it to be different for different methods? Generally speaking, I'm not big on using `NamedTuple` with optional fields as it's bound to lead to confusion and errors vs. explicit types with a few interface-methods implemented :confused:

Reply:
Yes I expect all objectives to maintain a state (it could be `nothing` for example.) So the fields of the top-most state object should be fixed. Should I use a struct to be more explicit?